debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/RecoverableConnectionHolder.java

```diff
@@ -0,0 +1,130 @@
+        this.user = user;
+        this.pwd = pwd;
+        this.waitRetryDelay = waitRetryDelay;
+        this.maxRetryCount = maxRetryCount;
+
+        createConnection();
+    }
+
+    private void createConnection() throws SQLException {
+        this.conn = DriverManager.getConnection(url, user, pwd);
+        this.conn.setAutoCommit(false);
+    }
+
+    @Override
+    public void close() throws SQLException {
+        conn.close();
+    }
+
+    public boolean isConnectionCreated() {
+        return conn != null;
+    }
+
+    public synchronized <T> T executeWithReconnect(ConnectionFunction<T> func, String name, boolean rollback)
+            throws SQLException, CustomException {
+        int attempt = 1;
+        while (true) {
+            if (conn == null) {
+                LOGGER.debug("Trying to reconnect (attempt {}).", attempt);
+                try {
+                    createConnection();
+                }
+                catch (SQLException e) {
+                    LOGGER.error("SQL Exception while trying to reconnect: " + e.getMessage(), e);
+                    conn = null;
+                    if (attempt >= maxRetryCount) {
+                        throw e;
+                    }
+                    attempt++;
+                    LOGGER.debug("Waiting for reconnect for {} ms.", waitRetryDelay);
+                    DelayStrategy delayStrategy = DelayStrategy.constant(waitRetryDelay);
+                    delayStrategy.sleepWhen(true);
+                    continue;
+                }
+            }
+            try {
+                return func.apply(conn);
+            }
+            catch (SQLRecoverableException e) {
+                LOGGER.warn("Attempt {} to call '{}' failed.", attempt, name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                try {
+                    conn.close();
+                }
+                catch (Exception ex) {
+                    // ignore
+                }
+                conn = null;
+
+            }
+            catch (SQLException e) {
+                LOGGER.warn("Call '{}' failed.", name, e);
+                if (rollback) {
+                    LOGGER.warn("'{}': doing rollback.", name);
+                    try {
+                        conn.rollback();
+                    }
+                    catch (SQLException ex) {
+                        // ignore
+                    }
+                }
+                throw e;
+            }
+        }
+    }
+
+    @FunctionalInterface
+    public interface ConnectionFunction<T> {
+        T apply(Connection conn) throws SQLException, CustomException;
+    }
+
+    public static class CustomException extends RuntimeException {
```

Comment:
Lets use `org.apache.kafka.connect.errors.RetriableException` instead of a new one.

Reply:
I opted for UncheckedIOException because the retriable exception would be really confusing here - it must not retry connection on it and it is used for further processing if action fails (see JdbcSchemaHistory.recoverRecords)